src/crafting.rs

```diff
@@ -110,55 +111,84 @@ fn calculate_precise_min_crafting_cost(
+    let output_item_count = recipe.map(|recipe| recipe.output_item_count).unwrap_or(1);
 
     let purchases_ptr = context.purchases.len();
     let crafted_ptr = context.crafted.len();
     let crafting_steps_before = context.crafting_steps;
 
+    // Take from leftovers first if any
+    let (item_count, cost_of_leftovers_used) = if let Some((count, cost)) = context.leftovers.remove(&item_id) {
+        match count.cmp(&item_count) {
+            std::cmp::Ordering::Less => {
+                // Source is only checked against crafting to break out of profit loop; so prefer
+                // whichever other source
+                context.leftovers.remove(&item_id);
+                (item_count - count, cost * count)
+            },
+            std::cmp::Ordering::Equal => {
+                context.leftovers.remove(&item_id);
+                return Some(PreciseCraftingCost {
+                    cost: cost * item_count,
+                    source: Source::Crafting,
+                })
+            }
+            std::cmp::Ordering::Greater => {
+                context.leftovers.insert(item_id, (count - item_count, cost));
+                return Some(PreciseCraftingCost {
+                    cost: cost * item_count,
+                    source: Source::Crafting,
+                })
+            }
+        }
+    } else {
+        (item_count, money::Money::zero())
+    };
+
+    // Craft x, but stash the rest; price is the fraction though
     let crafting_cost = recipe.and_then(|recipe| {
         if !opt.include_timegated && recipe.is_timegated() {
-            None
-        } else {
-            let mut cost = Rational32::zero();
-            for ingredient in &recipe.ingredients {
-                // adjust ingredient count based on fraction of parent recipe that was requested
-                let ingredient_count =
-                    Rational32::from(ingredient.count) * (item_count / output_item_count);
-                let ingredient_cost = calculate_precise_min_crafting_cost(
-                    ingredient.item_id,
-                    ingredient_count,
-                    recipes_map,
-                    items_map,
-                    tp_listings_map,
-                    context,
-                    opt,
-                );
-                if let Some(PreciseCraftingCost {
-                    cost: ingredient_cost,
-                    ..
-                }) = ingredient_cost
-                {
-                    cost += ingredient_cost;
-                } else {
-                    return None;
-                }
-            }
+            return None
+        }
 
-            Some(cost)
+        let mut cost = money::Money::zero();
+        for ingredient in &recipe.ingredients {
+            // adjust ingredient count based on fraction of parent recipe that was requested
+            let ingredient_count = Ratio::new(ingredient.count * item_count, output_item_count).ceil().to_integer();
```

Comment:
I think that `ingredient_count` should stay as a fraction. It handles cases where a recipe may need `x` amount of an ingredient, but the ingredient can only be crafted in batches of `y`. In those cases the results won't be the same as before.

Reply:
I assume `leftovers` is the replacement for it, but if keeping `ingredient_count` as a fraction allows `leftovers` to be removed then that would be better.

Reply:
One thing that frustrates me a little about the current method is that it doesn't list the leftovers from crafting - I'd like those accounted for.

In particular, it could be a net loss to make one more thing if it also requires making 5 (and using one) of a subcomponent which sells at a significant loss. I don't yet have code written to account for that (not quite sure if I will), but I'd like it to at least report the leftovers to me as part of the sales list along with price and total profit.

Reply:
It's a good point about the profit not being accurate if there are large outputs, and I agree that accounting for leftovers would be useful. The aim of using ratios in `calculate_precise_min_crafting_cost` was related to that issue, in order to not overestimate the cost of crafting either, when ingredients from one recipe can be shared between multiple items. We might be able to have the best of both worlds by repurposing `crafting_steps`:
https://github.com/t-mw/gw2-arbitrage/blob/3fe6c6929f72dcb5c94ce4544d66980baac59659/src/crafting.rs#L177

At the moment it equals the total number of recipes crafted, not distinguished by item. It could track the crafting steps per recipe instead e.g. `crafting_steps: HashMap<u32, Rational32>`. Once `calculate_precise_min_crafting_cost` is finished, if the fractional part of any of the ratios is > 0 then the final output of that recipe was not fully used and the fraction could be used to calculate what the actual remainder is.

Reply:
I've added vendor chunk purchasing and leftovers to the alt-currencies branch. Leftovers can come from crafting or purchasing.